docs/RFCS/20200414_scheduled_jobs.md

```diff
@@ -0,0 +1,383 @@
+  }
+
+  // ErrorHandlingBehavior describes how to handle failed job runs.
+  enum ErrorHandlingBehavior {
+    // By default, failed jobs will run again, based on their schedule.
+    RETRY_SCHED = 0;
+    // Retry failed jobs soon.
+    RETRY_SOON = 1;
+    // Stop running this schedule
+    PAUSE_SCHED = 2;
+  }
+
+  // The type of the executor to use.
+  ExecutorType executor_type = 1;
+
+  // How to handle running jobs.
+  WaitBehavior wait = 2;
+
+  // How to handle failed jobs.
+  ErrorHandlingBehavior on_error = 3;
+}
+```
+
+The default execution policy is:
+  * Wait for the outstanding job to finish before starting the next one
+    (`wait == WaitBehavior.WAIT`)
+  * If there were any errors in the execution, we continue attempting
+  to execute cron job (`on_error == RETRY_SCHED`)
+    
+We plan on supporting two types of executors for running periodic jobs:
+1. Inline executor: executes sql statement inline -- that is, scheduled job 
+   simply executes the job statement under the transaction.
+2. Async executor: execute sql statement as an external job.  Only the creation
+   of a job record performed inline.  The jobs system is responsible for the
+   actual execution of the statement.
+   
+For the V1, we will only support the `inline` executor.  The future versions will
+implement a job wrapper around arbitrary sql statement allowing end users to create
+custom cron jobs.  The async executor can also be used for running expensive
+statements as a job (for example `IMPORT INTO` -- those statements take long time
+to plan, and we may not want to do this under cron job execution transaction).
+
+In the subsequent iterations of this system, we can extend `ScheduleDetails`
+protocol messages to support more use cases.  Some future enhancements might include:
+  * Do not spawn new job if more than N instances of this job still running
+    (instead of restricting to 1)
+  * max_retries: Pause cron job if more than X runs resulted in an error.
+  * Add execution time limits (kill a job if it runs longer than X) 
+
+Finally, whenever we perform changes to the cron job (i.e. pause it), we want
+to record this information for debugging purposes in the following protocol message:
+
+```proto
+// ScheduleChangeInfo describes the reasons for schedule changes.
+message ScheduleChangeInfo {
+  message Change {
+    google.protobuf.Timestamp time = 1;
+    string reason = 2;
+  }
+  repeated Change changes = 1  [(gogoproto.nullable) = false];
+}
+```
+
+#### System database changes.
+Scheduled jobs stores job definitions in the `system.crontab` table:
+```sql
+CREATE TABLE system.crontab (
+    sched_id      INT DEFAULT unique_rowid() PRIMARY KEY,
+    job_name      STRING NOT NULL,
+    created       TIMESTAMPTZ NOT NULL DEFAULT NOW(),
+    schedule_expr STRING,
+    job_details   BYTES NOT NULL,
+    owner         STRING,
+    next_run      TIMESTAMPTZ,
+    exec_spec     BYTES,
+    change_info   BYTES,
+    INDEX         (next_run),
```

Comment:
I think we should family next_run separately since it will change frequently but everything else is expected to be pretty static.